File: src/moxfield.py
"""Moxfield API client for fetching deck and collection data."""

import re
from typing import Any, Dict, Optional

import requests

from .models import Card, Deck

MOXFIELD_API_BASE = "https://api2.moxfield.com/v2"
REQUEST_HEADERS = {
    "User-Agent": "MTGAssistant/1.0 (deck analysis tool)",
    "Accept": "application/json",
}


def extract_deck_id(url: str) -> Optional[str]:
    """Extract the public deck ID from a Moxfield URL.

    Supports formats:
    - https://www.moxfield.com/decks/AbCdEfGh
    - https://moxfield.com/decks/AbCdEfGh
    - AbCdEfGh  (raw ID passed directly)
    """
    url = url.strip()
    match = re.search(r"moxfield\.com/decks/([A-Za-z0-9_-]+)", url)
    if match:
        return match.group(1)
    # If it looks like a raw ID (no slashes, no spaces)
    if re.match(r"^[A-Za-z0-9_-]+$", url):
        return url
    return None


def get_deck(deck_id: str) -> Deck:
    """Fetch a deck from Moxfield by its public ID.

    Raises:
        requests.HTTPError: if the API returns a non-2xx status.
        ValueError: if the response cannot be parsed.
    """
    url = f"{MOXFIELD_API_BASE}/decks/all/{deck_id}"
    response = requests.get(url, headers=REQUEST_HEADERS, timeout=30)
    response.raise_for_status()
    return parse_moxfield_deck(response.json())


def parse_moxfield_deck(data: Dict[str, Any]) -> Deck:
    """Parse a Moxfield deck API response dict into a Deck object."""
    deck_id = data.get("publicId") or data.get("id", "")
    name = data.get("name", "Unknown Deck")

    # Commanders
    commanders_raw = data.get("commanders", {})
    commander_list = list(commanders_raw.values())
    commander = parse_moxfield_card(commander_list[0]) if len(commander_list) >= 1 else None
    partner = parse_moxfield_card(commander_list[1]) if len(commander_list) >= 2 else None

    # Mainboard
    mainboard = [
        parse_moxfield_card(entry) for entry in data.get("mainboard", {}).values()
    ]

    # Sideboard
    sideboard = [
        parse_moxfield_card(entry) for entry in data.get("sideboard", {}).values()
    ]

    return Deck(
        id=deck_id,
        name=name,
        commander=commander,
        partner=partner,
        mainboard=mainboard,
        sideboard=sideboard,
        format=data.get("format", "commander"),
        description=data.get("description", ""),
    )


def parse_moxfield_card(entry: Dict[str, Any]) -> Card:
    """Parse a single card entry from a Moxfield deck response."""
    quantity = entry.get("quantity", 1)
    card = entry.get("card", {})

    # Prefer the top-level image_uris; fall back to card_faces[0]
    image_uris = card.get("image_uris") or {}
    if not image_uris and card.get("card_faces"):
        image_uris = card["card_faces"][0].get("image_uris", {})

    oracle_text = card.get("oracle_text", "")
    if not oracle_text and card.get("card_faces"):
        oracle_text = card["card_faces"][0].get("oracle_text", "")

    return Card(
        name=card.get("name", ""),
        quantity=quantity,
        mana_cost=card.get("mana_cost", ""),
        cmc=float(card.get("cmc", 0)),
        type_line=card.get("type_line", ""),
        oracle_text=oracle_text,
        colors=card.get("colors", []),
        color_identity=card.get("color_identity", []),
        keywords=card.get("keywords", []),
        rarity=card.get("rarity", ""),
        set_code=card.get("set", ""),
        image_uri=image_uris.get("normal", ""),
        scryfall_id=card.get("scryfall_id", ""),
    )

File: supabase/functions/_shared/moxfield.ts
import { Card, createCard, Deck } from "./models.ts";

const MOXFIELD_API_BASE = "https://api2.moxfield.com/v2";

const BROWSER_HEADERS = {
  "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/134.0.0.0 Safari/537.36",
  Accept: "application/json, text/plain, */*",
  "Accept-Language": "en-US,en;q=0.9",
  Referer: "https://www.moxfield.com/",
  Origin: "https://www.moxfield.com",
};

export function extractDeckId(url: string): string | null {
  url = url.trim();
  const match = url.match(/moxfield\.com\/decks\/([A-Za-z0-9_-]+)/);
  if (match) return match[1];
  if (/^[A-Za-z0-9_-]+$/.test(url)) return url;
  return null;
}

export async function getDeck(deckId: string, timeoutMs = 10000): Promise<Deck> {
  const url = `${MOXFIELD_API_BASE}/decks/all/${encodeURIComponent(deckId)}`;
  
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), timeoutMs);
  
  try {
    const res = await fetch(url, { 
      headers: BROWSER_HEADERS,
      signal: controller.signal 
    });
    
    if (!res.ok) throw new Error(`Moxfield API error: ${res.status}`);
    const data = await res.json();
    return parseMoxfieldDeck(data);
  } catch (e) {
    if (e instanceof Error && e.name === 'AbortError') {
      throw new Error(`Moxfield API timeout after ${timeoutMs}ms`);
    }
    throw e;
  } finally {
    clearTimeout(timeout);
  }
}

export async function getDeckWithMeta(deckId: string, timeoutMs = 10000): Promise<[Deck, string]> {
  const url = `${MOXFIELD_API_BASE}/decks/all/${encodeURIComponent(deckId)}`;
  
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), timeoutMs);
  
  try {
    const res = await fetch(url, { 
      headers: BROWSER_HEADERS,
      signal: controller.signal 
    });
    
    if (!res.ok) throw new Error(`Moxfield API error: ${res.status}`);
    const data = await res.json();
    const deck = parseMoxfieldDeck(data);
    const lastUpdatedAt = data.lastUpdatedAtUtc || data.updatedAt || "";
    return [deck, lastUpdatedAt];
  } catch (e) {
    if (e instanceof Error && e.name === 'AbortError') {
      throw new Error(`Moxfield API timeout after ${timeoutMs}ms`);
    }
    throw e;
  } finally {
    clearTimeout(timeout);
  }
}

export function parseMoxfieldDeck(data: Record<string, unknown>): Deck {
  const deckId = (data.publicId as string) || (data.id as string) || "";
  const name = (data.name as string) || "Unknown Deck";

  const commandersRaw = (data.commanders || {}) as Record<string, unknown>;
  const commanderList = Object.values(commandersRaw);
  const commander = commanderList.length >= 1 ? parseMoxfieldCard(commanderList[0] as Record<string, unknown>) : null;
  const partner = commanderList.length >= 2 ? parseMoxfieldCard(commanderList[1] as Record<string, unknown>) : null;

  const mainboardRaw = (data.mainboard || {}) as Record<string, unknown>;
  const mainboard = Object.values(mainboardRaw).map((e) => parseMoxfieldCard(e as Record<string, unknown>));

  const sideboardRaw = (data.sideboard || {}) as Record<string, unknown>;
  const sideboard = Object.values(sideboardRaw).map((e) => parseMoxfieldCard(e as Record<string, unknown>));

  return {
    id: deckId,
    name,
    commander,
    partner,
    mainboard,
    sideboard,
    format: (data.format as string) || "commander",
    description: (data.description as string) || "",
  };
}

function parseMoxfieldCard(entry: Record<string, unknown>): Card {
  const quantity = (entry.quantity as number) || 1;
  const card = (entry.card || {}) as Record<string, unknown>;

  let imageUris = (card.image_uris || {}) as Record<string, string>;
  const cardFaces = (card.card_faces || []) as Record<string, unknown>[];
  if (!imageUris.normal && cardFaces.length > 0) {
    imageUris = (cardFaces[0].image_uris || {}) as Record<string, string>;
  }

  let oracleText = (card.oracle_text as string) || "";
  if (!oracleText && cardFaces.length > 0) {
    oracleText = (cardFaces[0].oracle_text as string) || "";
  }

  return createCard({
    name: (card.name as string) || "",
    quantity,
    mana_cost: (card.mana_cost as string) || "",
    cmc: Number(card.cmc || 0),
    type_line: (card.type_line as string) || "",
    oracle_text: oracleText,
    colors: (card.colors as string[]) || [],
    color_identity: (card.color_identity as string[]) || [],
    keywords: (card.keywords as string[]) || [],
    rarity: (card.rarity as string) || "",
    set_code: (card.set as string) || "",
    image_uri: imageUris.normal || "",
    scryfall_id: (card.scryfall_id as string) || "",
  });
}
